### ISSUE TITLE
L_PREFER_CANVAS and display vector object

### ISSUE THREAD

**Issue by Chrispnv** [mapbox/leaflet-image#82]
Hi,

I have web application with AngularJs, Leaflet 0.x.x and Postgres/Symfony2 server side for GeoJson vector data. I'm going to work to export part of map (tiles and vector) in pdf, so i use leaflet-image plugin. After a lot of tests, i have this issue.
- if i put L_PREFER_CANVAS = true; in script tag in my html file that manage all my application, polygons and linestring vector object are in canvas format (not markers) in map and when i click on button to fire to export image from map all objects are displayed good in image.
  ![capture](https://cloud.githubusercontent.com/assets/10419547/18000790/0a3e85c2-6b80-11e6-8b9e-dd68ababa4fe.PNG)
- if i put L_PREFER_CANVAS = true; in script tag in map html, polygons and linestring vector object are displayed good (i think vector object are not in canvas format) in map (markers also), but when i click on button to fire to export image from map, i have error message in chrome debug console :
  leaflet-image.js:185 Element could not be drawn on canvas <svg class=​" leaflet-zoom-animated" width=​"476" height=​"357" viewBox=​"-38 -29 476 357" style=​"transform:​ translate3d(-38px, -29px, 0px)​;​">​…​</svg>​handlePathRoot @ leaflet-image.js:185start @ leaflet-image.js:307poke @ leaflet-image.js:295(anonymous function) @ leaflet-image.js:323tileQueueFinish @ leaflet-image.js:162maybeNotify @ leaflet-image.js:348(anonymous function) @ leaflet-image.js:324im.onload @ leaflet-image.js:138 .....

Can you help me

Thanks.

Christophe

**Comment by tmcw** [comment]
Yep, L_PREFER_CANVAS must be set _before_ Leaflet or Mapbox.js is included, hence this problem.

**Comment by Chrispnv** [comment]
Thanks,
but why i have polygons and linestring vectors layers in block canvas and points layers in map.
I tried with simple example, that is all in html file, tile layer and polygon vector layer with L_PREFER_CANVAS=true. Polygon is displayed in map and also in image created by leaflet-image.
Do you have an idea on this issue ?
I'm going to simplify my code lines for not having vectors layers in canvas object on map.

Thanks
Christophe

**Comment by Chrispnv** [comment]
Hi,
I solved my problem. I had a canvas tag with attribute in my css file, therefore i removed this tag and it's ok.

Other problem resolved. My point vector are leaflet icon. In leaflet-image i had an error message : 'not read property 0 of undefined'. leaflet-image could not get iconSize and iconAnchor properties in handleMarkerLayer function. I added iconSize and iconAnchor properties in my leaflet icon (example : iconSize: [20, 20], iconAnchor: [10, 10])

Leaflet-image is a very good plugin allowing me to get image from map with IGN Geoportail tiles and vector layers (point, linestring and polygon), that will be cropped and inserted in PDF file with pdfmake.

Thanks
Christophe
